src/test/groovy/com/user_service/services/UserAuthenticationContractTest.groovy

```diff
@@ -0,0 +1,148 @@
+package com.user_service.services
+
+import com.user_service.controllers.UserAuthenticationController
+import com.user_service.dtos.request.*
+import com.user_service.dtos.response.*
+import org.springframework.http.HttpStatus
+import spock.lang.Specification
+
+import java.time.LocalDate
+
+class UserAuthenticationContractTest extends Specification {
+
+    UserAuthenticationController userAuthenticationController
+    UserAuthenticationService userAuthenticationService
+
+    def setup() {
+        userAuthenticationService = Mock(UserAuthenticationService)
+        userAuthenticationController = new UserAuthenticationController(userAuthenticationService)
+    }
+
+    def "Should register new user"() {
+        given: "SignUpUserDto is provided"
+            SignUpUserDto signUpUserDto =
+                    new SignUpUserDto("jarek", "jarek.jarkowski@gmail.com", LocalDate.of(1993, 7, 24), "dfG123!@0123456", "dfG123!@0123456")
+
+        and: "SignUpResponseDto is provided"
+            SignUpResponseDto signUpResponseDto =
+                    new SignUpResponseDto("j***@g***.com", "email", "1234")
+
+        and: "signUp method is mocked"
+            userAuthenticationService.signUpUser(signUpUserDto) >> signUpResponseDto
+
+        when: "signUp method is called"
+            def registerResponse = userAuthenticationController.signUpUser(signUpUserDto)
```

Comment:
Ideally, you should make an http call here. Calling a raw controller method is not exactly, how users will interact with this endpoint. We're skipping serialization part and some Spring magic here. Ideally the test should have structure:

```
given:
var request = 
    """
    { /json body here/
    """

when:
val response = restTemplate.postForEntity("localhost:8080", request)

then:
response.status == 201
response.getBody().email() == "j***@g***.com"
// other assertions
```

I simplified a bit request body JSON construction, don't remember exactly how it works there. Probably needs to be wrapped in some `HttpEntity` or sth like that. But our goal is to test through the whole stack, from HTTP request till the end. And mock as little as possible.